### PR TITLE
model: add tsinghua-ee/WAVE-7B omni embedding model

### PR DESCRIPTION
**Pull request by yaswanth169**
WAVE (arXiv:2509.21990) produces prompt-aware embeddings for text, audio, silent video, and synchronized audio-visual inputs, via a Qwen2.5-Omni-Thinker backbone with a BEATs audio encoder and an all-layer feature-fusion head.

The prior PR (#4841) stalled because WAVE's checkpoint shipped no modeling code. That's since been resolved upstream (trust_remote_code support merged into the HF repo), so this loads directly with no vendored copy of the training repo.

**Verification**

mteb mock-run, all four modalities pass:

| Modality | Pass |
|---|---|
| text | 41/41 |
| image | 15/15 |
| audio | 10/10 |
| video | 7/7 |

Real reproduction on AudioCapsA2TRetrieval (matches WAVE's paper benchmark "AudioCaps A-RET"), 883 audio queries against a 4411 item text corpus:

| Metric | Score |
|---|---|
| hit_rate@1 | 0.528 |
| hit_rate@5 (main) | 0.836 |
| hit_rate@10 | 0.913 |
| recall@100 | 0.958 |
| MRR@10 | 0.655 |

Not a strict reproduction of the paper's own protocol (different corpus size, different metric), reported as a health check rather than a paper exact number.

**Checklist**
- [x] ModelMeta filled out
- [x] mteb.get_model() and get_model_meta() work
- [x] Tested on a representative set of tasks
- [x] Model is public
- [x] Reproduced paper results, included above

### CODE REVIEW

**Comment by yaswanth169** [comment]
Hello @AdnanElAssadi56 and @Samoed , can you check this one now.

File: mteb/models/model_implementations/wave_models.py [hunk 1]
```diff
@@ -0,0 +1,484 @@
...
+            device_map=self.device,
+        )
+        # WAVE's own eval entrypoint reloads the BEATs submodule from the
+        # standalone checkpoint after `from_pretrained`, regardless of what
+        # `from_pretrained` already populated it with (see train_qwen.py's
+        # `run_test` path). Replicated here for a faithful reproduction.
+        # load_state_dict copies onto whatever device the target params are
+        # already on, so this works fine with the model already placed.
+        beats_ckpt = torch.load(beats_path, map_location="cpu")
+        self.model.beats.load_state_dict(beats_ckpt["model"])
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:160]: Shodn't be `to(device)` here?

File: mteb/models/model_implementations/wave_models.py [hunk 2]
```diff
@@ -0,0 +1,484 @@
...
+        width, height = image.size
+        if width < 28 or height < 28:
+            pad_width = max(0, 28 - width)
+            pad_height = max(0, 28 - height)
+            left, top = pad_width // 2, pad_height // 2
+            image = ImageOps.expand(
+                image,
+                border=(left, top, pad_width - left, pad_height - top),
+                fill=(0, 0, 0),
+            )
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:184]: I think processor inside do the same

File: mteb/models/model_implementations/wave_models.py [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+wave_7b = ModelMeta(
+    loader=Wave7BWrapper,
+    name="tsinghua-ee/WAVE-7B",
+    languages=["eng-Latn"],
+    open_weights=True,
+    revision="7d51cdaecfaabb9c529a447249cd4c2a6df8ce5b",
+    release_date="2026-02-11",
+    max_tokens=131_072,
+    n_parameters=9_410_651_007,
+    n_embedding_parameters=0,
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py]: This can't be none for this model

File: mteb/models/model_implementations/wave_models.py [hunk 4]
```diff
@@ -0,0 +1,484 @@
...
+        actual fps. Replicated here using the sampled frame count and the
+        source video's duration.
+        """
+        if frames.ndim == 4 and frames.shape[1] == 3 and frames.shape[-1] != 3:
+            frames = frames.permute(0, 2, 3, 1)
+        frames = frames.contiguous()
+        image_processor = self.processor.image_processor
+        video_proc = image_processor(images=None, videos=frames, return_tensors="pt")
+        fps = frames.shape[0] / duration if duration else self.fps
+        video_second_per_grid = [image_processor.temporal_patch_size / fps]
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py]: Why this needed to be done like this? Can you share source of this code?

File: mteb/models/model_implementations/wave_models.py [hunk 5]
```diff
@@ -0,0 +1,484 @@
...
+
+        feature_attention_masks = []
+        input_features = []
+        audio_length = 0
+        for segment in segments:
+            if segment.shape[0] < _SAMPLING_RATE:
+                segment = np.pad(  # noqa: PLW2901
+                    segment, (0, _SAMPLING_RATE - segment.shape[0])
+                )
+            features = feature_extractor(
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py]: Why not
```suggestion
            features = self.processor.feature_extractor(
```

File: mteb/models/model_implementations/wave_models.py [hunk 6]
```diff
@@ -0,0 +1,484 @@
...
+            video_proc["video_grid_thw"],
+            video_second_per_grid,
+        )
+
+    def _process_audio(
+        self, array: np.ndarray
+    ) -> tuple[dict[str, torch.Tensor], int, np.ndarray]:
+        """Mirror WAVE's process_audio: chunk into 300s segments for BEATs.
+
+        WAVE never truncates audio itself; long clips are chunked into 300s
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:228]: Maybe better just truncate to 300s?

File: mteb/models/model_implementations/wave_models.py [hunk 7]
```diff
@@ -0,0 +1,484 @@
...
+        from the nominal rate whenever the frame cap binds or a short video
+        keeps all its frames, and derives video_second_per_grid from that
+        actual fps. Replicated here using the sampled frame count and the
+        source video's duration.
+        """
+        if frames.ndim == 4 and frames.shape[1] == 3 and frames.shape[-1] != 3:
+            frames = frames.permute(0, 2, 3, 1)
+        frames = frames.contiguous()
+        image_processor = self.processor.image_processor
+        video_proc = image_processor(images=None, videos=frames, return_tensors="pt")
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py]:
```suggestion
        video_proc = self.processor.image_processor(videos=frames, return_tensors="pt")
```

File: mteb/models/model_implementations/wave_models.py [hunk 8]
```diff
@@ -0,0 +1,484 @@
...
+        text_input = self.processor.apply_chat_template(
+            conversation, add_generation_prompt=False, tokenize=False
+        )
+        text_input = text_input[0].split("<|im_start|>user\n")[-1].strip()
+
+        text_input = self.processor.replace_multimodal_special_tokens(
+            [text_input],
+            iter([audio_length]) if audio_inputs is not None else iter([]),
+            iter([grid_thw]) if grid_thw is not None else iter([]),
+            iter([video_grid_thw[0]]) if video_grid_thw is not None else iter([]),
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:326]: Why this need to be `iter`?

File: mteb/models/model_implementations/wave_models.py [hunk 9]
```diff
@@ -0,0 +1,484 @@
...
+            video_second_per_grid=iter(second_per_grid_ts)
+            if second_per_grid_ts is not None
+            else iter([]),
+            use_audio_in_video=audio_inputs is not None and video is not None,
+            position_id_per_seconds=25,
+            seconds_per_chunk=2.0 * second_per_grid_ts[0]
+            if second_per_grid_ts is not None
+            else None,
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:334]: Where did you find this?

**Comment by tomaarsen** [comment]
@yaswanth169 Have you also considered e.g. integrating the model with Sentence Transformers in the model repository itself? Then it can be more trivially tested, but also more trivially used. 

- Tom Aarsen

**Comment by yaswanth169** [comment]
@Samoed thanks for the review, addressed everything in the latest commit.

On the device question: no explicit .to(device) needed, self.model is already placed by device_map before this runs, and load_state_dict copies each tensor onto its target parameter's existing device regardless of what device the source checkpoint was loaded on.

On the image padding: you're right that Qwen2VLImageProcessor's own smart_resize already upscales images below its min_pixels bound, so this rarely fires on real inputs. I kept it since it's WAVE's own process_image_unified behavior and it's a harmless no-op when unneeded, cheap insurance for a min_pixels configuration where smart_resize's upscale alone wouldn't reach the 28px patch size floor.

On n_embedding_parameters: fixed, it's classify_linear's real parameter count, 372,513,792 (two Linear layers, hidden_size times num_hidden_layers to hidden_size, then hidden_size to hidden_size).

On the video fps source and the iter() question: both cited directly now. The fps derivation matches qwenvl/data/data_qwen.py's video_decord (line 274 in the WAVE repo), and position_id_per_seconds/seconds_per_chunk match process_omni_conversations in the same file (lines 449 to 450). The iter() wrapping is required because replace_multimodal_special_tokens, bundled in the checkpoint repo's processing_qwen2_5_omni.py, calls next() on those arguments internally.

On truncating audio to 300s instead of chunking: kept the chunking since it matches WAVE's own eval behavior exactly (they never truncate, only chunk and concatenate). In practice this rarely triggers since max_audio_length_seconds already caps most real inputs well under 300s.

On the two suggested diffs: applied both. One caught a real bug along the way, the image_processor(videos=frames) suggestion dropped the images=None argument, which is actually required by BaseImageProcessor.__call__() and crashed immediately on any video input. Fixed to keep images=None.

@tomaarsen thanks for the suggestion. Native Sentence Transformers integration would need changes in WAVE's own HF repo (or a full separate ST wrapper here), and WAVE's dual audio encoder plus all-layer fusion head don't map cleanly onto standard ST architectures. Open to revisiting if the WAVE authors add ST support upstream, but treating it as out of scope for this PR for now.

**Comment by yaswanth169** [comment]
Quick follow-up: re-verified these fixes on real GPU hardware. text, audio, image, and video all encode correctly end to end with proper shapes and normalized embeddings, no issues.

File: mteb/models/model_implementations/wave_models.py [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+            if audio_inputs is not None
+            else None,
+            "feature_attention_mask": audio_inputs["feature_attention_mask"]
+            if audio_inputs is not None
+            else None,
+            "input_raw_wav": [torch.from_numpy(raw_wav)]
+            if raw_wav is not None
+            else None,
+            "use_audio_in_video": video_grid_thw is not None
+            and audio_inputs is not None,
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:364]: Not possible to extract audio from our video
```suggestion
```

File: mteb/models/model_implementations/wave_models.py [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+        }
```
Comment (Samoed) [inline mteb/models/model_implementations/wave_models.py:365]: Is this really reqired? Why not just `processor.apply_chat_template`?

**Review by Samoed — COMMENTED** [review]
Can you reproduce scores from paper?

**Comment by tomaarsen** [comment]
> thanks for the suggestion. Native Sentence Transformers integration would need changes in WAVE's own HF repo (or a full separate ST wrapper here), and WAVE's dual audio encoder plus all-layer fusion head don't map cleanly onto standard ST architectures. Open to revisiting if the WAVE authors add ST support upstream, but treating it as out of scope for this PR for now.

Or a pull request, and the MTEB configs uses `revision="refs/pr/..."`.

- Tom Aarsen

**Comment by yaswanth169** [comment]
@Samoed ran real reproductions on the paper's own benchmarks instead of just the earlier health check (AudioCapsT2ARetrieval and ClothoT2ARetrieval, both already in mteb).

Results: AudioCaps recall@1 = 30.3% (paper reports R@1 = 44.2%), Clotho recall@1 = 20.7% (paper reports R@1 = 25.6%).
[wave_paper_repro_results.json](https://github.com/user-attachments/files/31947363/wave_paper_repro_results.json)

@tomaarsen thanks, the refs/pr/ pattern is good to know in general. It doesn't quite solve this specific case though: the blocker isn't where the ST config would live, it's that WAVE's dual audio encoders plus all-layer fusion head don't fit Sentence-Transformers' pooling-based single-tower interface. Even with an ST config PR upstream, mteb would still need the same custom encode logic this wrapper already has, so I don't think it reduces the integration work here. Happy to revisit if I'm missing something about how ST handles multi-encoder architectures.

**Comment by Samoed** [comment]
> Or a pull request, and the MTEB configs uses revision="refs/pr/...".

We can use commit hash from pr branch, but this can break after merge I think

> Results: AudioCaps recall@1 = 30.3% (paper reports R@1 = 44.2%), Clotho recall@1 = 20.7% (paper reports R@1 = 25.6%).

This seems very low. Can you investigate what is wrong?

**Comment by tomaarsen** [comment]
> but this can break after merge I think

That is possible, yeah. I'm not sure how that works. I know `refs/pr/...` does stay working, but it is obviously not fixed like a commit hash is.

And the WAVE model looks very custom, but it overall might be compatible with Sentence Transformers' multimodality. It's fairly flexible nowadays. Either way, I'm also fine with having implementations in MTEB, but there's a lot of value in having accessible implementations for each of these models. Otherwise we're back to like 2018 when a lot of research was essentially unusable, poorly accessible, and often stale. 

- Tom Aarsen

**Comment by yaswanth169** [comment]
@Samoed this run used 8-bit quantization, not full precision. No single GPU on the available hardware had the ~19GB free needed for bf16, and a multi-GPU split isn't viable for this model since WAVE's fusion head requires all layers on one device. That's a likely contributor to the gap versus the paper's numbers.

Haven't isolated exactly how much of the drop is quantization versus mteb's protocol differing from the paper's own eval script. Let me know if you'd like a genuine bf16 run to isolate it.

**Comment by Samoed** [comment]
Yes, would be good to see evaluation with full prescision

**Comment by ayush1298** [comment]
@yaswanth169 any update here?
